gradio/route_utils.py

```diff
@@ -583,3 +584,54 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+def get_core_url(original_url: str) -> str:
+    """
+    Returns the core url of a given url. If the url does not have a scheme, it is assumed to be http.
+    Examples:
+        get_core_url("https://www.gradio.app") -> "www.gradio.app"
+        get_core_url("localhost:7860") -> "localhost"
+        get_core_url("127.0.0.1") -> "127.0.0.1"
+    """
+    if not original_url:
+        return ""
+    if "://" not in original_url:
+        original_url = "http://" + original_url
+    return httpx.URL(original_url).host
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = request.headers.get("origin", "")
+        host_header_host = get_core_url(host_header)
+        origin_header_host = get_core_url(origin_header)
+
+        localhost_aliases = ["localhost", "127.0.0.1", "0.0.0.0", "null"]
```

Comment:
* Is `0.0.0.0` a localhost in this context? (https://superuser.com/questions/949428/whats-the-difference-between-127-0-0-1-and-0-0-0-0)
* What is `"null"`?

Reply:
Yes, `0.0.0.0` is localhost in this context. Basically, if a Gradio app is running on `0.0.0.0`, we should understand it to be running on a user's local machine and prevent third-party websites from making requests to it

Reply:
`null` is a bit of an interesting case, I discovered it experimentally. Basically it happens when you have a static webpage that you open in your browser. This can happen when you, e.g. embed a Gradio app, in an "index.html" file and then open that file in your browser.